### ISSUE TITLE
Phase 5: Attention head as decomposed graph (capstone composition validation)

### ISSUE THREAD

**Issue by Ravenwater** [branes-ai/graphs#127]
## Goal

Validate the **non-fused attention head** as a computational graph composed of operators already characterized in Phases 0-4. Capstone of the operator-expansion plan ([docs/v5/operator-validation-expansion-plan.md](../blob/main/docs/v5/operator-validation-expansion-plan.md)).

## Modeling stance: decomposition-first, no fusion

The attention head IS a computational graph of operators we already model. Phase 5 validates that running the analyzer over this graph produces a prediction matching the measured PyTorch eager implementation (which is exactly this graph, no fusion).

## Why decomposition-first is the right v1

* Fused-kernel reality (FlashAttention, FusedMHA) is a property of the deployment stack, not the hardware -- and the compiler/runtime layer is still under development.
* Before any compiler optimizations land, the realistic baseline is "PyTorch eager attention" -- exactly the unfused decomposition.
* Validating the decomposed graph proves the analyzer **composes**: per-op accuracy + correct dataflow plumbing = correct end-to-end prediction. If composed model is off, we know which sub-op to re-calibrate.
* When fused kernels do land (FlashAttention as a separate op in PR-D), measured-fused vs predicted-unfused = the **fusion uplift** -- an architecturally meaningful number that prioritizes compiler/runtime optimization work.

## The decomposition

Standard single-head attention as a graph of existing operators:

```
input: x ∈ (B, S, D_model)

  Q = linear(x, W_Q)           # Phase 0 (V4)
  K = linear(x, W_K)           # Phase 0 (V4)
  V = linear(x, W_V)           # Phase 0 (V4)
  scores = matmul(Q, K^T)      # Phase 0 (V4)
  scaled = elementwise(scores * 1/sqrt(D))  # Phase 1 / 2
  attn = softmax(scaled)       # Phase 3 (#125)
  out = matmul(attn, V)        # Phase 0 (V4)
  output = linear(out, W_O)    # Phase 0 (V4)
```

**Every node is an operator we already have**. There is no new "atomic" op to characterize.

## What the decomposition baseline does NOT model

This is honest about the gap:

* **FusedMHA / FlashAttention** -- avoid materializing the (B, H, S, S) scores by tiling Q and streaming K, V through L1. Measured fused latency can be 2-5x lower for long sequences.
* **Kernel-launch amortization** -- the decomposed graph pays N × launch overhead for N kernels (Q, K, V, scale, softmax, attn, output). Fused kernel pays one launch.
* **Inter-op reuse** -- scores tile that softmax reads is the SAME tile the attn matmul reads. Decomposed model double-counts; fused kernel keeps it in shared memory.

These gaps are **quantifiable once the baseline lands**. They become the prioritization signal for compiler / runtime optimization work.

## Variants to support

1. **Standard single-head** (B=1 or small) -- v1 deliverable.
2. **Multi-head batched** (split D_model into H heads, batch matmuls over H) -- v2.
3. **Grouped-query attention (GQA)** (Llama-style: K, V have fewer heads than Q) -- v3, lower priority.
4. **Causal masking** (~50% FLOPS on QK^T): skip in v1, add in v2.

## Footprint (single-head, decomposed, no fusion)

For input `(B, S, D_model)` and `D_head = D_model`:

| sub-op | FLOPS | bytes (no inter-op reuse) |
|---|---|---|
| Q = x·W_Q (linear) | 2*B*S*D² | 2*B*S*D + D² + 2*B*S*D |
| K = x·W_K (linear) | 2*B*S*D² | 2*B*S*D + D² + 2*B*S*D |
| V = x·W_V (linear) | 2*B*S*D² | 2*B*S*D + D² + 2*B*S*D |
| scores = Q·K^T (matmul) | 2*B*S²*D | 2*B*S*D + 2*B*S*D + 2*B*S² |
| scale (element-wise) | B*S² | 2*B*S² (in-place if measurer fuses) |
| softmax | ~5*B*S² | 2*B*S² (online) or 5*B*S² (naive) |
| attn = softmax·V (matmul) | 2*B*S²*D | 2*B*S² + 2*B*S*D + 2*B*S*D |
| out = attn·W_O (linear) | 2*B*S*D² | 2*B*S*D + D² + 2*B*S*D |

For typical configs (B=1, S=1024, D=128):
* FLOPS ≈ 8.6 GF
* Bytes ≈ 11 MB intermediates (scores S² dominates: 4 MB)
* On Orin Nano: scores S² (4 MB) doesn't fit L2 (2 MB) → DRAM round-trips

This is exactly the symptom FlashAttention solves -- the predicted latency will reflect that pain, providing the calibrated "before" for fusion gains.

## Deliverables (4 PRs)

* **PR-A**: Composite-op infrastructure (graph-walker so analyzer chains per-op `ReuseModel`s with inter-op dependencies / intermediate tensor lifetimes). Required before any composite predictions are meaningful.
* **PR-B**: Single-head attention baseline. Decomposition graph builder + V4 sweep generator over `(B, S, D)` + measurer + i7 + Orin Nano baselines.
* **PR-C**: Multi-head batched attention + causal masking variant.
* **PR-D**: FlashAttention as a separate op (optional, if PR-A's composition holds). Measured-fused vs predicted-unfused gives the quantified fusion uplift.

## Estimated scope

~7-14 days.

## Acceptance criteria

* [ ] Single-head attention predicts latency within 30% of measured (PyTorch eager) on i7 + Orin Nano for typical configs (S>=1024, D>=64)
* [ ] Multi-head batched attention predicts within 30%
* [ ] Composite-op infrastructure documented and unit-tested
* [ ] V4 floor tests pass for the new operator
* [ ] Comparison plot includes attention-head markers
* [ ] Documented gap between decomposed prediction and fused-kernel measurement (when FlashAttention measurements available)

## Dependencies

* **Blocks on**: Phase 3 (#125) -- composite-op infrastructure
* **Soft-depends on**: Phase 4 (#126) -- LayerNorm is part of full transformer block; attention head alone doesn't strictly need it

## Cross-link

* Plan: `docs/v5/operator-validation-expansion-plan.md` (Phase 5)
